### PR TITLE
Change github site-handler to use tokens for authorization

### PR DESCRIPTION
**Pull request by grembo**
Change github site-handler to make use of token based authorization using personal access tokens. This addresses https://github.com/freebsd/portscout/issues/8

Tried to clean up while there.

### CODE REVIEW

**Comment by koobs** [comment]
@grembo Change looks good. Do we have a token, or do we need to generate one?

**Comment by grembo** [comment]
> @grembo Change looks good. Do we have a token, or do we need to generate one?

We need to generate one from an account controlled by the project (I would assume that this is "freebsd", whoever is in charge of that).

The procedure is fairly easy:
- Go to https://github.com/settings/tokens [0]
- Click "Generate new token"
- Enter your password and click "Confirm password"
- Enter a fitting description in "Note:", e.g.
  "portscout access" or something more elaborate like
  "Check github upstreams of FreeBSD ports for new releases (portscout)"
- Do ***not*** add any scopes, as the purpose of this token is public read-only
- Click "Generate token"
- Copy and paste the token to portscout.conf (github token = _generatedtoken_)
  _Note: The token is only shown once._

You can see on https://github.com/settings/tokens if and when the token has been used. This allows you to verify that it was deployed successfully.

[0] Alternatively, click user icon -> Settings -> Developer settings -> Personal access tokens

**Comment by koobs** [comment]
> 
> 
> > @grembo Change looks good. Do we have a token, or do we need to generate one?
> 
> We need to generate one from an account controlled by the project (I would assume that this is "freebsd", whoever is in charge of that).
> 
> The procedure is fairly easy:
> 
>     * Go to https://github.com/settings/tokens [0]
> 
>     * Click "Generate new token"
> 
>     * Enter your password and click "Confirm password"
> 
>     * Enter a fitting description in "Note:", e.g.
>       "portscout access" or something more elaborate like
>       "Check github upstreams of FreeBSD ports for new releases (portscout)"
> 
>     * Do _**not**_ add any scopes, as the purpose of this token is public read-only
> 
>     * Click "Generate token"
> 
>     * Copy and paste the token to portscout.conf (github token = _generatedtoken_)
>       _Note: The token is only shown once._
> 
> 
> You can see on https://github.com/settings/tokens if and when the token has been used. This allows you to verify that it was deployed successfully.
> 
> [0] Alternatively, click user icon -> Settings -> Developer settings -> Personal access tokens

I'm on git-admin (and setup the freebsd org), so I can take care of that. We may want/need a portscout specific user to do this, rather than putting it under a user account. I'll investigate options and best-practice and relay findings here if I have questions

**Comment by grembo** [comment]
@koobs Just going through open pull requests I was wondering if this ever happened?

**Comment by koobs** [comment]
I think I may have and neglected to post them here. Will check again and generate a new set if I can't find.

**Comment by grembo** [comment]
> I think I may have and neglected to post them here. Will check again and generate a new set if I can't find.

Hm, but was the code change to make it work applied (as the pull request is still open) #confused

**Comment by koobs** [comment]
The changes have not been applied/deployed. Are you available on IRC to run through this and get it deployed?

File: Portscout/SiteHandler/GitHub.pm [hunk 1]
```diff
@@ -101,38 +101,22 @@ sub GetFiles
...
+	}
+	elsif ($url =~ /https:\/\/github\.com\/(.+?)\/archive\//) {
 		$projname = $1;
-	} elsif ($url =~ /https:\/\/github.com\/downloads\/(.+)\//) {
+	}
+	elsif ($url =~ /https:\/\/github.com\/downloads\/(.+)\//) {
 		$projname = $1;
-	} else {
+	}
+	else {
```
Comment (grembo) [inline Portscout/SiteHandler/GitHub.pm:111]: Seems like I just added newlines in there (for some reason)

File: Portscout/SiteHandler/GitHub.pm [hunk 2]
```diff
@@ -101,38 +101,22 @@ sub GetFiles
...
-
-	my $response = $ua->request(HTTP::Request->new(GET => $query));
-	if (!$response->is_success || $response->status_line !~ /^2/) {
-		_debug('GET failed: ' . $response->status_line);
-		return 0;
-	}
-
-	my $json = decode_json($response->decoded_content);
-	foreach my $release (@$json) {
+	my $releases = _call_github_api('/repos/' . $projname . '/releases')
```
Comment (grembo) [inline Portscout/SiteHandler/GitHub.pm:117]: Removed adding credentials to the URL (as those are done through a token in the new _call_github_api call)

File: Portscout/SiteHandler/GitHub.pm [hunk 3]
```diff
@@ -101,38 +101,22 @@ sub GetFiles
...
+	  or return 0;
+	foreach my $release (@$releases) {
```
Comment (grembo) [inline Portscout/SiteHandler/GitHub.pm:119]: Now we deal with releases an not json here (as decoding was done in the _call_github_api  call)

File: Portscout/SiteHandler/GitHub.pm [hunk 4]
```diff
@@ -141,20 +125,9 @@ sub GetFiles
...
-
-		if (!$response->is_success || $response->status_line !~ /^2/) {
-			_debug('GET failed: ' . $response->status_line);
-			return 0;
-		}
-		$json = decode_json($response->decoded_content);
-		foreach my $tag (@$json) {
+		my $tags = _call_github_api('/repos/' . $projname . '/tags')
+		  or return 0;
+		foreach my $tag (@$tags) {
```
Comment (grembo) [inline Portscout/SiteHandler/GitHub.pm:130]: Same as above (thanks to having the function call, we won't have to repeat everything).

File: portscout.conf [hunk 5]
```diff
@@ -175,8 +175,10 @@ db connstr       = DBI:Pg:dbname=%(db_name)
 # GitHub site handler settings
 # GitHub rate limits requests to its API to a very low number for unauthenticated
 # requests, and 5000 per hour for authenticated requests.
+# GitHub personal access tokens can be requested on github accounts that
+# have a verified email address here: https://github.com/settings/tokens
+# A public personal access token without any special permissions will do(!)
 
-# github_client_id	=				# GitHub Client ID
-# github_client_secret	=				# GitHub Client Secret
+#github token     =				# GitHub personal access token
```
Comment (grembo) [inline portscout.conf:182]: Remove client_id/client_secret from code and replace it by github token (which really needs an underscore in here).

File: Portscout/SiteHandler/GitHub.pm [hunk 6]
```diff
@@ -165,6 +138,41 @@ sub GetFiles
...
+#------------------------------------------------------------------------------
+# Func: _call_github_api()
+# Desc: Calls the github api making use of settings.
+#
+# Args: $resource - Resource to query (e.g. "/repos/project/releases")
+#
+# Retn: Parsed JSON
+#------------------------------------------------------------------------------
+
+sub _call_github_api {
```
Comment (grembo) [inline Portscout/SiteHandler/GitHub.pm:150]: This function encapsulates calls to github. It does GET requests to api.github.com, adds _github_token_ if one was configured and decodes the JSON format response.

**Review by grembo — COMMENTED** [review]
So two things to change:

- Remove newline only changes, as they make things less obvious
- Add underscore to sample config (github_token).

File: Portscout/SiteHandler/GitHub.pm (hunk 1 above)
Comment (grembo) [reply Portscout/SiteHandler/GitHub.pm]: Reverted that part of the change

File: Portscout/SiteHandler/GitHub.pm (hunk 3 above)
Comment (koobs) [reply Portscout/SiteHandler/GitHub.pm:116]: Note: Also uses tags below

File: Portscout/SiteHandler/GitHub.pm (hunk 6 above)
Comment (koobs) [reply Portscout/SiteHandler/GitHub.pm:147]: Nice!

**Review by koobs — APPROVED** [review]
My perl foo aint super, but this LGTM, and this has been tested on a local install, which we'll repeat pre deployment to production

**Pull request merged**